Introduce dispatchable shape function

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Proposed in #20549

#### Brief description of what is fixed or changed

`shape()` is introduced to get the shape of object representing matrix.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- tensor
    - `shape()` function is introduced in array module.
<!-- END RELEASE NOTES -->

JSS95 commented:
@oscarbenjamin @Upabjojr 

I opened this PR to implement `shape()` function, which was proposed in #20549. I expect that we may need different handlers for expressions with same class but different kinds, so I introduced `PropertyDispatcher` class. `shape()` function will be instance of `PropertyDispatcher`.

Although I didn't actually introduce `shape()` in 87f8361, docstring of `PropertyDispatcher` shows the overall plan for it. Please take a look and let me know your idea.

Upabjojr commented on sympy/core/kind.py:
```diff
@@ -385,3 +386,94 @@ def __doc__(self):
         return '\n\n'.join(docs)
+
+
+class PropertyDispatcher:
```
Do you think it's wise to introduce new ad-hoc dispatchers? I'm pointing this out because it could make analyzing the code a bit harder.

We have already multipledispatching, can we use that in order to dispatch on both `kind` and `expr`?

As an alternative, can't we have a single-dispatcher over the kinds, returning a shape getter?
```python
def shape(expr):
    get_shape = shape_of_kind(expr.kind)
    return get_shape(expr)

@shape_of_kind.register
def _(kind: NumberKind):
    return lambda expr: ()

@shape_of_kind.register
def _(kind: ArrayKind):
    return get_shape_for_array

@get_shape_for_array.register
def _(expr: Integral)
    return shape(expr.args[0])
```

Another possibility would be using the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern)
```python
def shape(expr):
    return accept_shape(expr.kind, expr)

@accept_shape.register
def _(kind: NumberKind, expr):
    return visit_shape_numberkind(expr)

@accept_shape.register
def _(kind: ArrayKind, expr)
    return visit_shape_arraykind(expr)

@visit_shape_numberkind.register
def _(expr: Integral):
    return ()  # or None or NotImplementedError() if you prefer

@visit_shape_arraykind.register
def _(expr: Integral):
    return shape(expr.args[0])
```

The second approach has the advantage of being translatable into C++, which is theory could be used to port the code to [SymEngine](https://github.com/symengine/symengine) if we're going to replace parts of SymPy with C++ code in the future.

JSS95 replied:
Thanks. I was uneasy with `PropertyDispatcher`, and this looks better.

oscarbenjamin commented:
For most objects we have that either the object has a shape attribute or its shape is `shape(obj.args[0])`. Likewise most classes that don't have a unique `kind` will have `kind(obj) == kind(obj.args[0])`. I wonder if there's a way to specify this so that both `kind` and `shape` can use the same information.

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
+
+def shape(expr):
+    """
+    Return the shape of the tensor.
```
I think that it would be better to say "Return the shape of the matrix or tensor as a tuple"

oscarbenjamin replied:
Actually I think maybe `array` is better than `tensor` here.

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
+    Return the shape of the tensor.
+
+    This function dispatches over the kind and class of *expr* to return
+    its shape.
```
If `shape` is an end user function then this should not be mentioned at the top of the docstring.

Perhaps in a "Notes" section at the bottom there could be a brief mention that `accept_shape` is used as the mechanism for implementing this function and can be used to extend it to define the shape of other kinds of objects so that `shape()`. There can be a "See also" referencing `accept_shape` and then the `accept_shape` docstring can explain how to overload shape for a new kind and how to implement e.g. matrix shape for a new class.

oscarbenjamin replied:
What is worth mentioning is the contrast between `shape(obj)` and `obj.shape` and the fact that `shape(obj)` applies in a more general context. We should make sure that there are "See also" referencing in both directions from `.shape` and `shape()`.

Probably the matrices tutorial should be updated to use `shape()` or should at least mention it.

oscarbenjamin commented on sympy/matrices/common.py:
```diff
@@ -2989,6 +2991,46 @@ def __iter__(self): # supports numpy.matrix and numpy.array
+    See Also
+    ========
+
+    sympy.tensor.ArrayKind : Kind for N-dimensional arrays.
```
Could be a see also here for `shape` and also an example showing how to use it.

It's probably worth mentioning that `shape(obj)` is expected to work whenever the `kind` is `MatrixKind`. It is the key property that is implemented for this `kind`.

JSS95 replied:
We may have more functions like `shape` (for example, we may need `rank` function) in the future, and it will be unnecessary to mention every functions in each `Kind`'s docstrings. I think mentioning `shape()` in tutorial doc will suffice.

oscarbenjamin replied:
Somewhere there should be a list of the basic things that you can do with objects of a given `Kind`. In general we should have a lot more "See also" cross-referencing in the docs because it is very useful for someone reading the docs who does not already know the codebase.

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
+
+@singledispatch
+def accept_shape(kind, expr):
+    raise NotImplementedError
```
I'm not sure about the name `accept_shape`. Maybe `get_shape_for_kind`?

I don't think `NotImplementedError` is right. This will be raised when someone calls `shape` on something invalid like a `Set`. I guess `TypeError` would be better but actually the best would be a dedicated error class that can be documented and caught. It could be `NoShapeError`.

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
+
+@accept_shape.register(_NumberKind)
+def _(kind, expr):
+    return ()
```
Is it useful to have an empty tuple as shape for a scalar?

I would expect it to be more useful to get an error since this will usually be a bug.

oscarbenjamin commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -560,3 +562,15 @@ def as_immutable(self):
+
+@accept_shape.register(ArrayKind)
+def _(kind, expr):
+    return visit_shape_arraykind(expr)
```
Rather than `visit_shape_arraykind` it would be better to call this something like `get_shape_array` because this function gets the shape of an array.

JSS95 commented:
I wonder if it will be better to make `shape()` dispatch over class first. Such as:

```python
@singledispatch
def shape(expr):
    return get_shape_for_kind(expr.kind, expr)

@shape.register(Integral)
def _(expr):
    return shape(expr.args[0])
```

This way we don't need to manually register `Integral` to `get_shape_matrix`, `get_shape_array`, etc.

oscarbenjamin commented on sympy/matrices/common.py:
```diff
@@ -2989,6 +2991,46 @@ def __iter__(self): # supports numpy.matrix and numpy.array
+    Examples
+    ========
+
+    >>> from sympy import MatrixSymbol, Integral
+    >>> from sympy.abc import x
+    >>> A = MatrixSymbol('A', 2,2)
+    >>> A.kind
+    MatrixKind(NumberKind)
+    >>> Integral(A,x).kind
+    MatrixKind(NumberKind)
```
I think this should be expanded to something that makes it clear what the difference is between checking `isinstance(obj, Matrix)` and checking `obj.kind == MatrixKind(NumberKind)`. There should be text explaining when and why you should check the kind. Also how would you check for `MatrixKind` without specifying the element kind?

oscarbenjamin commented on sympy/matrices/common.py:
```diff
@@ -2989,6 +2991,46 @@ def __iter__(self): # supports numpy.matrix and numpy.array
+
+    element_kind : Kind
+        Kind of the element. Default is ``NumberKind``, which means that
+        the matrix contains only numbers.
```
Can NumberKind be a linked cross-reference in the docs?

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +62,116 @@ def __getitem__(self, index):
+
+def shape(expr):
+    """
+    Return the shape of the array as a tuple.
```
Matrices are a more common use in sympy than arrays so the word matrix should be here in the first line of this function. The vast majority of the time this will be used to get the shape of a matrix and not an array.

I know that you have talked about unifying matrices and arrays but they are not unified now. We should write docs now that correctly specify the current behaviour. (SymPy already has docs written based on hopeful future changes to the codebase and they are highly problematic)

oscarbenjamin replied:
Maybe this docstring should have a parameters section and a raises section that references NoShapeError. The NoShapeError docstring should show how to import and catch that error.

oscarbenjamin commented on sympy/tensor/functions.py:
```diff
@@ -61,3 +62,116 @@ def __getitem__(self, index):
+
+    Generally, this returns another singledispatch function such as
+    :obj:`get_shape_array() <sympy.tensor.array.ndim_array.get_shape_array>`
+    or :obj:`get_shape_matrix() <sympy.matrices.common.get_shape_matrix>`.
```
Can these references be
```
:py:func:`~.get_shape_array`
```

JSS95 commented:
I modified the docstrings. BTW @oscarbenjamin @Upabjojr  please let me know how do you think of making `shape()` as singledispatch function, as I mentioned in my previous comment:

> I wonder if it will be better to make shape() dispatch over class first. Such as:
> ```
> @singledispatch
> def shape(expr):
>     return get_shape_for_kind(expr.kind, expr)
> 
> @shape.register(Integral)
> def _(expr):
>     return shape(expr.args[0])
> ```
> This way we don't need to manually register Integral to get_shape_matrix, get_shape_array, etc.

Upabjojr commented:
> @Upabjojr please let me know how do you think of making `shape()` as singledispatch function, as I mentioned in my previous comment:

It works for me, it looks simpler. I would recommend to make the second function (i.e. `get_shape_for_kind`) private by prepending `_` to its name.

JSS95 commented:
I refactored the structure, but didn't make `get_shape_for_kind` private yet. I would like to know @oscarbenjamin 's thought first.

Upabjojr commented:
> I refactored the structure, but didn't make `get_shape_for_kind` private yet. I would like to know @oscarbenjamin 's thought first.

If you find it easier, you can also use `@dispatch` from `sympy.multipledispatch`, which can dispatch on all function parameters (and not just the first one, as `@singledispatch` currently does). Unfortunately it's not standard Python, but you can define a doubly-dispatched function in one line (without the need of the two single-dispatched functions).

oscarbenjamin commented:
I wonder if dispatching on kind and then class is overkill. The implementation is the same for both ArrayKind and MatrixKind as far as I can tell:

1. If it's an integral then return `shape(obj.args[0])`.
2. If it's a matrix/array class then return `obj.shape`.

Can we not just do something like:
```python
def shape(obj):
    if obj.kind not in (MatrixKind, ArrayKind):
        raise TypeError("shape is only defined for matrices and arrays")
    return _get_shape(obj)

@singledispatch
def _get_shape(obj):
     raise NotImplementedError("shape is not implemented for %s" % type(obj))

@_get_shape.register(Matrix)
@_get_shape.register(MatExpr)
@_get_shape.register(Array)
...
def _(obj):
    return obj.shape

@_get_shape.register(Integral)
def _(obj):
    return shape(obj.function)
```

JSS95 commented:
I want to make `shape` dispatchable over other kinds. For example I want to define `TensorKind` in my library and make `sympy.shape` return its shape. 
```pythin
if obj.kind not in (MatrixKind, ArrayKind):
````
 will make it impossible - is there any way around?

oscarbenjamin commented:
The simplest way would be something like `if obj.kind.has_shape` or perhaps something like:
```
def shape(obj):
    get_shape = getattr(obj.kind, 'get_shape', None)
    if get_shape is None:
        raise TypeError
    else:
        return get_shape(obj)
```
Then `MatrixKind` and `ArrayKind` could share a `get_shape` function if they don't need to have different implementations. Any new Kind is free to add a `get_shape` function if it wants to overload `shape`.

If there are useful extensions of `shape` then perhaps the current way does make sense. I think that multiple/single dispatch is not something that we should reach for every time we need to dispatch though. If nothing else there is a runtime cost for each dispatch so doing a double dispatch every time we get the shape of a matrix might be a measurable slowdown for some operations. Note that getting the shape is typically done as a check at the start of functions that operate on matrices.

JSS95 commented:
Maybe we can remove kind checking from `shape()` to avoid slowdown.

```python
@singledispatch
def shape(expr):
    if hasattr(expr, "shape"):
        return expr.shape
    raise TypeError("%s does not have shape, or not registered to shape() function." % expr)

@shape.register(Integral)
def _(expr):
    return shape(expr.function)
```

JSS95 commented:
@Upabjojr @oscarbenjamin I removed kind checking in `shape()`.

oscarbenjamin commented:
This looks good to me

JSS95 commented:
Then I will squash the commits and wait for @Upabjojr to review.

JSS95 committed:
feat(core): Introduce dispatchable shape function

Introduce shape function to tensor/functions.py
Register Integral to shape
Move MatrixKind from matrices.py to common.py
